src/Components/Selection.re

```diff
@@ -0,0 +1,75 @@
+[@deriving show({with_path: false})]
+type t = {
+  anchor: int,
+  focus: int,
+};
+
+type directionVariant =
+  | Left(int)
+  | Right(int);
+
+let initial: t = {anchor: 0, focus: 0};
+
+let withinLength = (text: string, position: int): int => {
+  let withingStart = max(position, 0);
+  let withingBoth = min(withingStart, String.length(text));
+
+  withingBoth;
+};
+
+let create = (~text: string, ~anchor: int, ~focus: int): t => {
+  let safeAnchor = withinLength(text, anchor);
+  let safeFocus = withinLength(text, focus);
+
+  {anchor: safeAnchor, focus: safeFocus};
+};
+
+let anchor = (selection: t): int => {
+  selection.anchor;
+};
+
+let focus = (selection: t): int => {
+  selection.focus;
+};
+
+let range = (selection: t): int => {
+  abs(selection.focus - selection.anchor);
+};
+
+let rangeStart = (selection: t): int => {
+  min(selection.focus, selection.anchor);
+};
+
+let rangeEnd = (selection: t): int => {
+  max(selection.focus, selection.anchor);
+};
```

Comment:
It's not clear to me what these functions mean, even after seeing their implementation. What is a "range" here? To me a "range" implies at least a start and an end, i.e. two numbers, or it could be mean the elements between the start and end, but I can't understand how a single number can be a range unless one of the ends are implied.